### PR TITLE
Check if GAE application is on cloud infrastructure

### PR DESCRIPTION
**Pull request by mdrachuk**
This PR adds the `ServerEnvironment#isProductionAppEngine()` flag and updates the `ServerEnvironment` internals to smoothly handle more properties based on `SystemProperty`.

### CODE REVIEW

**Comment by mdrachuk** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/ServerEnvironment.java [hunk 1]
```diff
@@ -44,33 +56,57 @@ private ServerEnvironment() {
...
+    @VisibleForTesting
+    static void resetProvider() {
+        provider.set(SystemEnvironmentProvider.newInstance());
+    }
+
+    /**
+     * The provider of the deployment type.
+     */
+    @Internal
+    public interface Provider {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/ServerEnvironment.java]: Let's use JDK `Supplier` instead.

File: server/src/main/proto/spine/server/server_environment.proto [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.server";
+option java_outer_classname = "ServerEnvironmentProto";
+option java_multiple_files = true;
+
+// A type of the environment system is deployed to.
+//
+enum DeploymentType {
+
+    // Running on a unrecognized environment.
```
Comment (armiol) [inline server/src/main/proto/spine/server/server_environment.proto]: We run in a standalone mode. Meaning that no specific runtime deployment is recognized.

File: server/src/main/proto/spine/server/server_environment.proto [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    STANDALONE = 0;
+
+    // Running on an App Engine local development server.
+    APPENGINE_EMULATOR = 2;
```
Comment (armiol) [inline server/src/main/proto/spine/server/server_environment.proto]: Please fix the indexes.

File: server/src/main/proto/spine/server/server_environment.proto [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+option java_outer_classname = "ServerEnvironmentProto";
+option java_multiple_files = true;
+
+// A type of the environment system is deployed to.
+//
+enum DeploymentType {
+
+    // Running in a standalone mode.
+    //
+    // Standalone mode is activated when no specific runtime deployment is recognized.
```
Comment (armiol) [inline server/src/main/proto/spine/server/server_environment.proto]: `The deployment is considered `STANDALONE` when no specific ...`

File: server/src/main/java/io/spine/server/SystemEnvironmentSupplier.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.function.Supplier;
+
+import static com.google.common.base.Strings.emptyToNull;
+import static io.spine.server.DeploymentType.APPENGINE_CLOUD;
+import static io.spine.server.DeploymentType.APPENGINE_EMULATOR;
+import static io.spine.server.DeploymentType.STANDALONE;
+import static java.util.Optional.ofNullable;
+
+/**
+ * The Default implementation of {@linkplain io.spine.server.ServerEnvironment.Provider
```
Comment (armiol) [inline server/src/main/java/io/spine/server/SystemEnvironmentSupplier.java]: `default`.

File: server/src/main/java/io/spine/server/SystemEnvironmentSupplier.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ * server environment provider}.
+ */
+class SystemEnvironmentSupplier implements Supplier<DeploymentType> {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/SystemEnvironmentSupplier.java]: `DeploymentDetector`, as this class does not supply `SystemEnvironment`.

File: server/src/main/java/io/spine/server/ServerEnvironment.java [hunk 7]
```diff
@@ -24,17 +24,29 @@
...
+    private static final String ENV_KEY_APP_ENGINE_RUNTIME_VERSION =
+            "com.google.appengine.runtime.version";
 
     /** If set, contains the version of AppEngine obtained from the system property. */
+    @SuppressWarnings("AccessOfSystemProperties") /*  Based on system property. */
     private static final @Nullable String appEngineRuntimeVersion =
-            System.getProperty(ENV_KEY_APP_ENGINE_RUNTIME_VERSION);
+            emptyToNull(System.getProperty(ENV_KEY_APP_ENGINE_RUNTIME_VERSION));
+
+    private static final ThreadLocal<DeploymentType> provider =
```
Comment (armiol) [inline server/src/main/java/io/spine/server/ServerEnvironment.java]: The deployment type is the same for all threads. So using `ThreadLocal` doesn't make much sense.

**Pull request merged**